### PR TITLE
dm/optimistic: make errors which isn't sync error handled by 'binlog skip' command

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/3789

### What is changed and how it works?
Add ConflictStage `ConflictError` and returns an error directly when dm-worker meets this stage.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has persistent data change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to update the documentation

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/tests/shardddl3_1/run.sh [hunk 1]
```diff
@@ -532,16 +532,23 @@ function DM_117_CASE {
...
+		"binlog-schema update test ${shardddl1} ${tb1} ${WORK_DIR}/schema.sql -s mysql-replica-01" \
+		"\"result\": true" 2
+
+	# skip this error
 	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
-		'shard-ddl-lock unlock "test-`shardddl`.`tb`" -s mysql-replica-01 --action exec -d shardddl1 -t tb1' \
-		"\"result\": true" 1
+		"binlog skip test" \
+		"\"result\": true" 2 \
+		"\"source 'mysql-replica-02' has no error\"" 1
```
Comment (GMHDBJD) [inline dm/tests/shardddl3_1/run.sh:551]: How about check the task status after skip? Maybe insert some new rows.

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM

**Comment by GMHDBJD** [comment]
chaos test https://github.com/pingcap/tiflow/actions/runs/1979146851

File: dm/tests/shardddl3_1/run.sh (hunk 1 above)
Comment (lichunzhu) [reply dm/tests/shardddl3_1/run.sh:551]: https://github.com/pingcap/tiflow/pull/4883/commits/02270f79808d88bd6b2d0af0e1fa3843cc9af326

**Comment by Ehco1996** [comment]
will review after codefreeze 😂

**Comment by lichunzhu** [comment]
https://ci.pingcap.net/blue/rest/organizations/jenkins/pipelines/ut-check/runs/5762/nodes/45/steps/49/log/?start=0
Should be fixed after https://github.com/pingcap/tiflow/pull/4893

**Comment by lichunzhu** [comment]
@Ehco1996 @GMHDBJD PTAL

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: dm/syncer/optimist.go [hunk 2]
```diff
@@ -179,6 +179,9 @@ func (s *Syncer) handleQueryEventOptimistic(qec *queryEventContext) error {
 		}
 	}
 
+	if op.ConflictStage == optimism.ConflictError {
```
Comment (Ehco1996) [inline dm/syncer/optimist.go:182]: maybe use switch to handle different `op`, ok to change in next pr

**Review by Ehco1996 — APPROVED** [review]
LGTM

**Pull request merged**